Drop support for EOL Python 2.7

hugovk opened the pull request:
For #105.

Intentionally leaves out dropping the `PendingDeprecationWarning`, so they can be done in another PR, and also doesn't drop 3.5 yet to keep the PR a bit smaller.

florisla commented:
Looks good!

Pylint complains about three minor things:

    bumpversion\cli.py:114:0: C0301: Line too long (122/100) (line-too-long)
    bumpversion\cli.py:620:0: C0301: Line too long (109/100) (line-too-long)
    bumpversion\version_part.py:109:12: R1724: Unnecessary "elif" after "continue" (no-else-continue)

Do we still need too support the third-party `configparser` package?
https://pypi.org/project/configparser/

I'm not sure what the story was on that, but it seems like we're currently not importing that anywhere.  Yet there are separate AppVeyor and Tox steps to test against it.

Based on a simple search, there are still some Python-2-isms in

    bumpversion\tests\test_cli.py:24
    bumpversion\utils.py:11
    bumpversion\README.md:153
    bumpversion\README.md:187
    bumpversion\README.md:276
    bumpversion\README.md:290
    bumpversion\README.md:317
    bumpversion\README.md:342

ekohl commented:
> I'm not sure what the story was on that, but it seems like we're currently not importing that anywhere. Yet there are separate AppVeyor and Tox steps to test against it.

It's probably a left over from old Python versions and ok to drop.

hugovk commented:
> Looks good!
> 
> Pylint complains about three minor things:
> 
> ```
> bumpversion\cli.py:114:0: C0301: Line too long (122/100) (line-too-long)
> bumpversion\cli.py:620:0: C0301: Line too long (109/100) (line-too-long)
> bumpversion\version_part.py:109:12: R1724: Unnecessary "elif" after "continue" (no-else-continue)
> ```

Fixed.

> Based on a simple search, there are still some Python-2-isms in
> 
> ```
> bumpversion\tests\test_cli.py:24
> bumpversion\utils.py:11
> bumpversion\README.md:153
> bumpversion\README.md:187
> bumpversion\README.md:276
> bumpversion\README.md:290
> bumpversion\README.md:317
> bumpversion\README.md:342
> ```

Removed, except I wasn't sure what the `utils.py` one is.

florisla commented:
My bad; I thought DiscardDefaultIfSpecifiedAppendAction was added to work around a bug which got fixed in the end -- but it's not been modified.

If you want to do further cleanup, there are still some `from __future__` imports, like `print_function` that can be dropped.

hugovk commented:
Already removed the `__future__` imports! The future is now!

florisla commented:
RIP Python 2, let's merge this!

ekohl commented:
Should we squash this on merge into a single commit? This now adds 3.8 support as well, but that's also in master.

florisla commented:
@ekohl You could fix that as merge conflicts, no?  Or else rebase on master first.

tobiasherp commented:
I'm **against** dropping 2.7 support _just because of the official end of life._ I'm quite sure there are many developers out there (including me) who would love to switch to Python 3 but still can't because they have a big project to maintain which includes a lot of packages which are not yet completely py3 compatible. In my case, I have a quite complex Plone 4.3 setup to maintain, and it will take me quite a while to update to Plone 5.2 which supports Python 3.

If you feel you must -- please do it in a new branch, and why not bump the major version to 1, since this package is travis-testable at last?

ekohl commented:
> If you feel you must -- please do it in a new branch, and why not bump the major version to 1, since this package is travis-testable at last?

This is IMHO a major version bump, but master shouldn't be held back. At most a python2 branch should be created.

hugovk commented:
The current releases which work on Python 2 will still be installable from PyPI and won't go away. 

And this project uses `python_requires` so modern pip will install the correct library version for your running Python:

* https://github.com/c4urself/bump2version/blob/master/setup.py

florisla commented:
@tobiasherp Asking out of curiosity -- do you have Python 3 available at all on your OSes?

It's possible to install bump2version in another Python than the one you use for development or hosting.

And would you have a need for recent `bump2version` features in such older projects?

tobiasherp commented:
OK, so be it.

c4urself commented:
agree with @hugovk -- there's no real reason to not drop support when folks can use an older version.

c4urself commented on README.md:
```diff
@@ -27,7 +27,7 @@ commits and tags:
 * works without any VCS, but happily reads tag information from and writes
   commits and tags to Git and Mercurial if available
 * just handles text files, so it's not specific to any programming language
-* supports Python2, Python3 and Pypy
```
Let's add a separate notice that if folks want to use Python2 they can target >1.0.0

hugovk replied:
If people have modern pip 9+, the `python_requires='>=3.5'` metadata will ensure they get the last supported version.

So in https://github.com/c4urself/bump2version/pull/108/commits/32933285b70a6c24fc77106ac8a532fa952f4de5 I added:

> If you want to use Python 2, use `pip>=9` and you'll get the last supported version, or pin `bump2version<1`.

How's that sound? (And I assume you meant `less than` 1.0.0 :)

c4urself replied:
yep, my bad :)

c4urself commented on bumpversion/cli.py:
```diff
@@ -11,19 +8,18 @@
 import sre_constants
 import sys
 import warnings
+from configparser import (
```
let's put imports in alphabetical order

hugovk replied:
Done! https://github.com/c4urself/bump2version/pull/108/commits/1eb846bb7b6c74955262840273b93b744c21341b

c4urself commented on tests/test_cli.py:
```diff
@@ -1,33 +1,27 @@
 import platform
 import warnings
 import subprocess
+from configparser import RawConfigParser
```
kind of a nit: but let's make import alphabetical, it makes it easier to determine duplicative imports etc.

hugovk replied:
Done! https://github.com/c4urself/bump2version/pull/108/commits/1eb846bb7b6c74955262840273b93b744c21341b

c4urself commented on tox.ini:
```diff
@@ -1,20 +1,18 @@
 [tox]
-envlist = py27, py27-configparser, py35, py36, py37, py38, pypy, pypy3
+envlist = py35, py36, py37, py38, py38-configparser, pypy3
```
the original reason to install `configparser` instead of using the default library's was that the py3 configparser was different (and preferred). Now that we no longer need Python2 we also do not need the third-party configparser, let's remove those options

hugovk replied:
How does https://github.com/c4urself/bump2version/pull/108/commits/ff9fec876b9fb1b69fc7d2fa790d8fa2bf303467 look?

ekohl commented on README.md:
```diff
@@ -27,7 +27,12 @@ commits and tags:
+* supports Python 3 and PyPy3
+
+If you want to use Python 2, use `pip>=9` and you'll get the last supported version,
+or pin `bump2version<1`.
```
I'd phrase it slightly different:
```suggestion
When using Python 2, pip 9.0.0 and higher will install a compatible version using [PEP440's](https://www.python.org/dev/peps/pep-0440/) [python_requires](https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires). With older pip versions a manual pin to `bump2version < 1` is needed.
```

hugovk replied:
Looks good, please could you add a space to "PEP 440" in the suggestions? That's how they're written:

https://www.python.org/dev/peps/pep-0440/

Plus add code formatting backticks around `python_requires`?

c4urself commented:
thanks @hugovk

hugovk commented:
You're welcome!
